repo/tuf_on_ci/client.py

```diff
@@ -5,74 +5,110 @@
+    print(f"Role {role} is valid on {ref_time}: OK")
+
+
 @click.command()
 @click.option("-v", "--verbose", count=True, default=0)
 @click.option("-m", "--metadata-url", type=str, required=True)
 @click.option("-a", "--artifact-url", type=str, required=True)
+@click.option("-u", "--update-base-url", type=str)
 @click.option("-e", "--expected-artifact", type=str)
-@click.option(
-    "-c",
-    "--metadata-cache",
-    type=str,
-    help="Optional client cache dir that client should use",
-)
+@click.option("--compare-source/--no-compare-source", default=True)
+@click.option("-t", "--time", type=int)
+@click.option("-o", "--offline-time", type=int)
+@click.option("-d", "--metadata-dir", type=str)
 def client(
     verbose: int,
     metadata_url: str,
     artifact_url: str,
+    update_base_url: str | None,
     expected_artifact: str | None,
-    metadata_cache: str | None,
+    compare_source: bool,
+    time: int | None,
+    offline_time: int | None,
+    metadata_dir: str | None,
 ) -> None:
-    """Test client for tuf-on-ci
-
-    Client expects to be run in tuf-on-ci repository (current metadata
-    in the sources will be considered the expected expected metadata the
-    client should receive from remote)."""
+    """Test client for tuf-on-ci"""
 
     logging.basicConfig(level=logging.WARNING - verbose * 10)
 
     with TemporaryDirectory() as client_dir:
-        metadata_dir = os.path.join(client_dir, "metadata")
+        if metadata_dir is None:
+            metadata_dir = os.path.join(client_dir, "metadata")
         artifact_dir = os.path.join(client_dir, "artifacts")
-        os.mkdir(metadata_dir)
+        os.makedirs(metadata_dir, exist_ok=True)
         os.mkdir(artifact_dir)
 
-        if metadata_cache:
-            # pre-populate the client cache with given directory
-            for fname in os.listdir(metadata_cache):
-                fpath = os.path.join(metadata_cache, fname)
-                if os.path.isfile(fpath):
-                    shutil.copy(fpath, os.path.join(metadata_dir, fname))
-        else:
-            # pre-populate the client cache with first root version from source
-            root = "metadata/root_history/1.root.json"
-            shutil.copy(root, os.path.join(metadata_dir, "root.json"))
-
-        # For now, just confirm we can get top level metadata from remote
+        # initialize client with a root.json from metadata_url
+        root_url = f"{metadata_url}/1.root.json"
+        try:
+            request.urlretrieve(root_url, f"{metadata_dir}/root.json")  # noqa: S310
+        except OSError as e:
+            sys.exit(f"Failed to download initial root {root_url}: {e}")
+
+        if update_base_url is not None:
+            # Update client to update_base_url before doing the actual update
+            updater = Updater(metadata_dir, metadata_url, artifact_dir, artifact_url)
+            try:
+                updater.refresh()
```

Comment:
Is the intended purpose of `update_base_url` to only verify that it can update with the new repo, and then continue below to perform the tests against the configured url.

I'm thinking if this needs clarification when e.g. both `update_base_url` and `validity` is provided, is it clear which repo the time validity is tested against?

Reply:
Wait, I think i understand now. This is only used to load the cache on disk, then below on line 77 we update again to get the fresh metadata, which we actually test again, right?

Reply:
Yes that sounds 100% correct:
* metadata_url is always what were really testing: so `valid_days` etc are always checked on data from that URL
* normally we only ensure that the refresh works if local cache contains 1.root.json
* if update_base_url is specified we first refresh from that URL (to populate local cache), then ensure refresh works: this makes sure e.g. snapshot contents don't illegally change between the two repos

Reply:
Perfect!

Reply:
oh wait you found a bug -- it uses the wrong URL :joy:

Reply:
oh yes, it should be `update_base_url`.

Reply:
fixed now, thanks!

Reply:
it's not verifiable in the action output but with debug output once can see that the correct url is getting used:

`tuf-on-ci-test-client -vvv -m https://sigstore.github.io/root-signing-staging/ -a https://tuf-repo-cdn.sigstage.dev/targets/  --no-compare-source --update-base-url https://tuf-repo-cdn.sigstage.dev/

`